faiss/gpu/impl/RaftIVFPQ.cu

```diff
@@ -0,0 +1,546 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <faiss/gpu/GpuIndexFlat.h>
+#include <faiss/gpu/utils/RaftUtils.h>
+#include <faiss/gpu/impl/FlatIndex.cuh>
+#include <faiss/gpu/impl/RaftIVFPQ.cuh>
+#include <faiss/gpu/utils/Transpose.cuh>
+
+#include <raft/neighbors/ivf_pq.cuh>
+#include <raft/neighbors/ivf_pq_helpers.cuh>
+
+#include <limits>
+#include <memory>
+
+namespace faiss {
+namespace gpu {
+
+RaftIVFPQ::RaftIVFPQ(
+        GpuResources* resources,
+        int dim,
+        idx_t nlist,
+        faiss::MetricType metric,
+        float metricArg,
+        int numSubQuantizers,
+        int bitsPerSubQuantizer,
+        bool useFloat16LookupTables,
+        bool useMMCodeDistance,
+        bool interleavedLayout,
+        float* pqCentroidData,
+        IndicesOptions indicesOptions,
+        MemorySpace space)
+        : IVFPQ(resources,
+                dim,
+                nlist,
+                metric,
+                metricArg,
+                numSubQuantizers,
+                bitsPerSubQuantizer,
+                useFloat16LookupTables,
+                useMMCodeDistance,
+                interleavedLayout,
+                // skip ptr allocations in base class (handled by RAFT
+                // internally) false,
+                pqCentroidData,
+                indicesOptions,
+                space) {
+    FAISS_THROW_IF_NOT_MSG(
+            indicesOptions == INDICES_64_BIT,
+            "only INDICES_64_BIT is supported for RAFT index");
+}
+
+RaftIVFPQ::~RaftIVFPQ() {}
+
+void RaftIVFPQ::reserveMemory(idx_t numVecs) {
+    fprintf(stderr,
+            "WARN: reserveMemory is NOP. Pre-allocation of IVF lists is not supported with RAFT enabled.\n");
+}
+
+void RaftIVFPQ::reset() {
+    raft_knn_index.reset();
+}
+
+size_t RaftIVFPQ::reclaimMemory() {
+    fprintf(stderr,
+            "WARN: reclaimMemory is NOP. reclaimMemory is not supported with RAFT enabled.\n");
+    return 0;
+}
+
+void RaftIVFPQ::setPrecomputedCodes(Index* quantizer, bool enable) {}
+
+idx_t RaftIVFPQ::getListLength(idx_t listId) const {
+    FAISS_ASSERT(raft_knn_index.has_value());
+    const raft::device_resources& raft_handle =
+            resources_->getRaftHandleCurrentDevice();
+
+    uint32_t size;
+    raft::update_host(
+            &size,
+            raft_knn_index.value().list_sizes().data_handle() + listId,
+            1,
+            raft_handle.get_stream());
+    raft_handle.sync_stream();
```

Comment:
Done.

Reply:
I had to undo this change because the `raft_knn_index->lists()->size.load()` does not give the exact size, whereas the list_sizes array in raft contains the exact sizes. As a result, the former was giving segmentation faults.

Reply:
During search only the `list_size `array is used, so it is correct to use that value.

Still `list[i].size` shall be the same as `list_size[i]`. When we resize the list, then we set the `size `member variable to the new size. We should check once more where the difference comes from. This need not block this PR, I have added a task to #3249